File: web/build-search-index.js
const fs = require("fs");
const path = require("path");
const glob = require("path");
const IR_DIR = path.join("public", "data", "ir");
const SEARCH_DIR = path.join("public", "data", "search");
const NUM_SEARCH_DIR = path.join("public", "data", "search-num");
const PSC_SEARCH_DIR = path.join("public", "data", "search-psc");

function cleanPrefix(name) {
  const chars = [];
  for (const ch of name.toLowerCase()) {
    if (/[a-z]/.test(ch)) {
      chars.push(ch);
      if (chars.length === 2) break;
    }
  }
  return chars.length < 2 ? "zz" : chars.join("");
}

function numPrefix(number) {
  const clean = (number || "").toUpperCase().replace(/[^A-Z0-9]/g, "");
  return clean.length >= 2 ? clean.slice(0, 2) : "00";
}

function pscPrefix(name) {
  const skip = new Set(["mr","mrs","ms","miss","dr","sir","lord","lady","prof","professor","dame"]);
  const words = name.toLowerCase().replace(/[^a-z\s]/g, "").trim().split(/\s+/).filter(w => !skip.has(w));
  const surname = words.length > 0 ? words[words.length - 1] : "";
  const chars = [];
  for (const ch of surname) {
    if (/[a-z]/.test(ch)) {
      chars.push(ch);
      if (chars.length === 2) break;
    }
  }
  return chars.length < 2 ? "zz" : chars.join("");
}

function main() {
  if (!fs.existsSync(SEARCH_DIR)) fs.mkdirSync(SEARCH_DIR, { recursive: true });
  if (!fs.existsSync(NUM_SEARCH_DIR)) fs.mkdirSync(NUM_SEARCH_DIR, { recursive: true });
  if (!fs.existsSync(PSC_SEARCH_DIR)) fs.mkdirSync(PSC_SEARCH_DIR, { recursive: true });

  const files = fs.readdirSync(IR_DIR).filter(f => f.endsWith(".json")).sort();
  console.log(`Building search index from ${files.length} IR files...`);

  const index = {};
  const numIndex = {};
  const pscIndex = {};
  const pscSeen = {};
  let total = 0;
  let pscTotal = 0;

  for (let i = 0; i < files.length; i++) {
    const fp = path.join(IR_DIR, files[i]);
    const data = JSON.parse(fs.readFileSync(fp, "utf-8"));
    const industry = data.industry;
    const industryName = data.industryName;
    const region = data.region;
    const regionName = data.regionName;

    for (const c of data.companies || []) {
      const name = c.name || "";
      const number = c.number || "";
      const town = c.postTown || "";
      if (!name) continue;

      const entry = { n: name, no: number, t: town, i: industry, in: industryName, r: region, rn: regionName };

      // Name index
      const prefix = cleanPrefix(name);
      if (!index[prefix]) index[prefix] = [];
      index[prefix].push(entry);

      // Number index
      if (number) {
        const np = numPrefix(number);
        if (!numIndex[np]) numIndex[np] = [];
        numIndex[np].push(entry);
      }

      // PSC index
      if (c.pscs && c.pscs.length > 0) {
        for (const psc of c.pscs) {
          const pscName = psc.n || "";
          if (!pscName) continue;

          const dedupeKey = `${pscName}|||${number}`;
          if (pscSeen[dedupeKey]) continue;
          pscSeen[dedupeKey] = true;

          const pp = pscPrefix(pscName);
          const pscEntry = {
            pn: pscName,
            pt: psc.t,
            n: name,
            no: number,
            t: town,
            i: industry,
            in: industryName,
            r: region,
            rn: regionName
          };
          if (!pscIndex[pp]) pscIndex[pp] = [];
          pscIndex[pp].push(pscEntry);
          pscTotal++;
        }
      }

      total++;
    }

    if ((i + 1) % 50 === 0) {
      console.log(`  Processed ${i + 1}/${files.length} files (${total.toLocaleString()} companies, ${pscTotal.toLocaleString()} PSC entries)...`);
    }
  }

  console.log(`\n  Total: ${total.toLocaleString()} companies across ${Object.keys(index).length} prefix groups`);
  console.log(`  Number index: ${Object.keys(numIndex).length} prefix groups`);
  console.log(`  PSC index: ${pscTotal.toLocaleString()} entries across ${Object.keys(pscIndex).length} prefix groups`);
  console.log(`  Writing search index...`);

  for (const [prefix, entries] of Object.entries(index)) {
    fs.writeFileSync(path.join(SEARCH_DIR, `${prefix}.json`), JSON.stringify(entries));
  }

  for (const [prefix, entries] of Object.entries(numIndex)) {
    fs.writeFileSync(path.join(NUM_SEARCH_DIR, `${prefix}.json`), JSON.stringify(entries));
  }

  for (const [prefix, entries] of Object.entries(pscIndex)) {
    fs.writeFileSync(path.join(PSC_SEARCH_DIR, `${prefix}.json`), JSON.stringify(entries));
  }

  console.log(`  Done! Search index built.`);
}

main();

File: web/build-search-index.py
#!/usr/bin/env python3
"""
Build a search index for the front page.
Splits 2.6M companies into ~676 files by first 2 chars of name.
Each file is small enough to fetch on demand.

Run from the web folder:
    python build-search-index.py
"""
import json, os, glob, re

IR_DIR = os.path.join("public", "data", "ir")
SEARCH_DIR = os.path.join("public", "data", "search")

def clean_prefix(name):
    """Get first 2 lowercase alpha chars from company name."""
    chars = []
    for ch in name.lower():
        if ch.isalpha():
            chars.append(ch)
            if len(chars) == 2:
                break
    if len(chars) < 2:
        return "zz"  # fallback for numeric/weird names
    return "".join(chars)

def main():
    os.makedirs(SEARCH_DIR, exist_ok=True)
    
    files = sorted(glob.glob(os.path.join(IR_DIR, "*.json")))
    print(f"Building search index from {len(files)} IR files...")
    
    # Collect all companies
    index = {}  # prefix -> list of entries
    total = 0
    
    for i, fp in enumerate(files, 1):
        with open(fp, "r", encoding="utf-8") as f:
            data = json.load(f)
        
        industry = data["industry"]
        industryName = data["industryName"]
        region = data["region"]
        regionName = data["regionName"]
        
        for c in data.get("companies", []):
            name = c.get("name", "")
            number = c.get("number", "")
            town = c.get("postTown", "")
            
            if not name:
                continue
            
            prefix = clean_prefix(name)
            
            entry = {
                "n": name,           # company name
                "no": number,        # company number
                "t": town,           # town
                "i": industry,       # industry slug
                "in": industryName,  # industry name
                "r": region,         # region slug
                "rn": regionName,    # region name
            }
            
            if prefix not in index:
                index[prefix] = []
            index[prefix].append(entry)
            total += 1
        
        if i % 50 == 0:
            print(f"  Processed {i}/{len(files)} files ({total:,} companies)...")
    
    print(f"\n  Total: {total:,} companies across {len(index)} prefix groups")
    
    # Also build a number index (by first 2 chars of company number)
    # This goes into the same files with a "num_" prefix
    
    # Write index files
    print(f"\n  Writing search index files to {SEARCH_DIR}...")
    
    # Write a manifest of all prefixes and their sizes
    manifest = {}
    
    for prefix, entries in sorted(index.items()):
        outpath = os.path.join(SEARCH_DIR, f"{prefix}.json")
        with open(outpath, "w", encoding="utf-8") as f:
            json.dump(entries, f, ensure_ascii=False, separators=(",", ":"))
        
        size_kb = os.path.getsize(outpath) / 1024
        manifest[prefix] = {"count": len(entries), "sizeKB": round(size_kb, 1)}
    
    # Write manifest
    manifest_path = os.path.join(SEARCH_DIR, "_manifest.json")
    with open(manifest_path, "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2)
    
    # Stats
    sizes = [v["sizeKB"] for v in manifest.values()]
    total_mb = sum(sizes) / 1024
    print(f"\n  Index files: {len(manifest)}")
    print(f"  Total size: {total_mb:.1f} MB")
    print(f"  Avg file: {sum(sizes)/len(sizes):.1f} KB")
    print(f"  Largest: {max(sizes):.1f} KB")
    print(f"  Smallest: {min(sizes):.1f} KB")
    
    print(f"\nDone! Search index built in {SEARCH_DIR}/")
    print(f"Add 'public/data/search/' to .gitignore (it's generated at build time)")

if __name__ == "__main__":
    main()
